server/src/tram_track_graph_transformer/tram_track_graph_transformer.py

```diff
@@ -277,7 +277,35 @@ def densify_graph_by_max_distance(
                     path_nodes[-1],
                 )
 
-        if errors:
-            raise ExceptionGroup("Track direction errors during densification", errors)
+        # if errors:
+        #     raise ExceptionGroup("Track direction errors during densification", errors)
 
         return densified_graph
+
+    def get_node_by_id(self, id: int) -> Node | None:
+        """
+        Returns node with the given ID if it exists
+
+        Parameters
+        ----------
+        id : int
+
+        Returns
+        ----------
+        Node or None
+        """
+        if id in self._nodes_by_id:
+            return self._nodes_by_id[id]
+        return None
+
+    def get_permament_nodes(self) -> set[Node]:
+        """
+        Get set of permament nodes from the graph
+        """
+        return self._permanent_nodes
```

Comment:
I'm not sure when, but I think the `permanent_nodes` property has been removed from the `TramTrackGraphTransformer` class. You can use the `@property` decorator in order to achieve a better interface:
```suggestion
    @property
    def permament_nodes(self):
        return self._permanent_nodes.copy()
```
I'm not a huge fan of return type hints where it's not necessary, as the IDE should figure it out for you anyways and this way we'll avoid mistakes with changing the return type but not the type hint.

I'm also not a big fan of *obvious* docstrings, I believe that `permanent_nodes` contains enough information to know what this property contains.